### PR TITLE
[twitch] Switch access token to GraphQL and refactor.

### PR DESCRIPTION
**Pull request by 23rd**
### Before submitting a *pull request* make sure you have:
- [ ] [Searched](https://github.com/ytdl-org/youtube-dl/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [ ] Read [adding new extractor tutorial](https://github.com/ytdl-org/youtube-dl#adding-support-for-a-new-site)
- [x] Read [youtube-dl coding conventions](https://github.com/ytdl-org/youtube-dl#youtube-dl-coding-conventions) and adjusted the code to meet them
- [ ] Covered the code with tests (note that PRs without tests will be REJECTED)
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

### In order to be accepted and merged into youtube-dl each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Bug fix
- [x] Improvement
- [ ] New extractor
- [ ] New feature

---

### Description of your *pull request* and other information

The main purpose of this PR is switching endpoint for the access token from the old `api.twitch.tv/api` to the newest GraphQL.
Looks like the `api.twitch.tv/api` endpoint is still working, but it will be rejected at any time. The `twitch.tv` frontend already uses GraphQL requests to get the access token.

Side changes:
- <del> Removed unused `_download_access_token` method.</del>
- Moved `TwitchGraphQLBaseIE` class to the top after `TwitchBaseIE` class.
- Added a new `_download_base_gql` method to make general GraphQL requests and remove repeated `https://gql.twitch.tv/gql` endpoint strings.

I am open to suggestions for improving the code of this PR.

### CODE REVIEW

File: youtube_dl/extractor/twitch.py [hunk 1]
```diff
@@ -151,16 +151,71 @@ def _prefer_source(self, formats):
...
+
+    def _download_access_token_gql(self, video_id, method, param_name, note):
+        ops = {
+            'query': '''{
+              %s(
+                %s: \"%s\",
+                  params: {
+                    platform: \"web\",
+                    playerBackend: \"mediaplayer\",
+                    playerType: \"site\"
```
Comment (dstftw) [inline youtube_dl/extractor/twitch.py:198]: Do not escape quotes inside triple quotes.

**Review by dstftw — CHANGES_REQUESTED** [review]

File: youtube_dl/extractor/twitch.py [hunk 2]
```diff
@@ -442,8 +465,8 @@ def _entries(self, channel_name, *args):
 
     # Deprecated kraken v5 API
     def _entries_kraken(self, channel_name, broadcast_type, sort):
-        access_token = self._download_access_token(channel_name)
-        channel_id = self._extract_channel_id(access_token['token'], channel_name)
+        access_token = self._download_stream_access_token_gql(channel_name)
+        channel_id = self._extract_channel_id(access_token['value'], channel_name)
```
[the remark this answers is not in the document]
Comment (23rd) [reply youtube_dl/extractor/twitch.py]: Hmm, okay…

File: youtube_dl/extractor/twitch.py [hunk 3]
```diff
@@ -160,7 +160,67 @@ def _extract_channel_id(self, token, channel_name):
...
+
+    def _download_vod_access_token_gql(self, video_id):
+        return self._download_access_token_gql(
+            video_id, 'videoPlaybackAccessToken', 'id',
+            'Downloading %s access token JSON GraphQL' % video_id)
+
+    def _download_stream_access_token_gql(self, channel_name):
+        return self._download_access_token_gql(
+            channel_name, 'streamPlaybackAccessToken', 'channelName',
+            'Downloading access token JSON GraphQL')
```
Comment (dstftw) [inline youtube_dl/extractor/twitch.py:220]: Inline.

Comment (dstftw) [inline youtube_dl/extractor/twitch.py:220]: 215, 220 DRY.

File: youtube_dl/extractor/twitch.py [hunk 4]
```diff
@@ -160,7 +160,67 @@ def _extract_channel_id(self, token, channel_name):
...
+                signature
+              }
+            }''' % (method, param_name, video_id),
+        }
+        return self._download_base_gql(video_id, ops, note)['data'][method]
+
+    def _download_vod_access_token_gql(self, video_id):
+        return self._download_access_token_gql(
+            video_id, 'videoPlaybackAccessToken', 'id',
+            'Downloading %s access token JSON GraphQL' % video_id)
```
Comment (dstftw) [inline youtube_dl/extractor/twitch.py:215]: -`JSON`.

File: youtube_dl/extractor/twitch.py (hunk 3 above)
Comment (dstftw) [inline youtube_dl/extractor/twitch.py:220]: -`JSON`.

**Review by dstftw — CHANGES_REQUESTED** [review]

**Comment by galipmedia** [comment]
Can anyone here help me? I am trying to hit the endpoint as specified in the code but I am not getting an access token back as the 'value' but rather a json object...
```
  data: {
    videoPlaybackAccessToken: {
      value: '{"authorization":{"forbidden":false,"reason":""},"chansub":{"restricted_bitrates":[]},"device_id":null,"expires":1610841132,"https_required":true,"privileged":false,"user_id":null,"version":2,"vod_id":1234567}',
      signature: '99aecf120103b0bb375bad907aa17a880c51d3c0',
      __typename: 'PlaybackAccessToken'
    }
  },
  extensions: { durationMilliseconds: 70, requestID: '01EW4N4TJSA64ACG62Q2G6KPZM' }
}
```
I am using the same Client-ID that was working with the old api.

**Comment by TaoziDB** [comment]
'value' you got is a string, the same as 'value' from using old API.

**Comment by galipmedia** [comment]
Aha thanks @TaoziDB it works, i just had to uri encode it for the next call.

**Pull request closed without merging**